### PR TITLE
ZEPPELIN-2687: Show code in description section of Spark UI

### PR DESCRIPTION
**Pull request by karuppayya**
### What is this PR for?
Show code in description section of Spark UI

### What type of PR is it?
 Improvement

### Todos
* [ ] - Task

### What is the Jira issue?
ZEPPELIN-2687

### How should this be tested?
Run scala, pyspark and sql snippet(should result in a spark job), and check Spark Web UI

### Screenshots (if appropriate)
Before:
screen shot 2017-06-24 at 12 26 02

After:
Collapsed(short desc):
screen shot 2017-06-24 at 12 12 38

Expanded(Long desc):
screen shot 2017-06-24 at 12 12 43

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

File: spark/src/main/java/org/apache/zeppelin/spark/Utils.java [hunk 1]
```diff
@@ -135,4 +135,33 @@ public static String getUserName(AuthenticationInfo info) {
     }
     return uName;
   }
+
+  public static String getJobShortText(String paraText) {
```
Comment (zjffdu) [inline spark/src/main/java/org/apache/zeppelin/spark/Utils.java]: Could you add unit test for these util methods ?

File: spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java [hunk 2]
```diff
@@ -1183,6 +1184,9 @@ public InterpreterResult interpret(String line, InterpreterContext context) {
   public InterpreterResult interpret(String[] lines, InterpreterContext context) {
     synchronized (this) {
       z.setGui(context.getGui());
+      String paraText = context.getParagraphText();
+      sc.setCallSite(new CallSite(Utils.getJobShortText(paraText),
```
Comment (zjffdu) [inline spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java]: `SparkRInterpreter` is missing here

**Comment by zjffdu** [comment]
@karuppayya Thanks for this PR. It is really helpful. Besides this, with more and more features added we have more and more duplicated code in `PySparkInterpreter`, `SparkInterpreter`, `SparkSQLInterpreter` and `SparkRInterpreter`.  It is better to have a base class to handle all these common things. Anyway, we could do this in another PR. But I just feel we need to to asap. Because zeppelin code base become too large than it could be.

**Comment by karuppayya** [comment]
@zjffdu +1 on the refactoring.

File: spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java (hunk 2 above)
Comment (karuppayya) [inline spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java]: @zjffdu I didnot find api to `setLocalProperty()` in R.    
 Any idea/pointer on how to implement his with sparkR?

Comment (zjffdu) [inline spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java]: hmm, then it is fine not do it for SparkR for now. Could you create a spark ticket for that ?

Comment (karuppayya) [inline spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java]: filed https://issues.apache.org/jira/browse/SPARK-21208

File: spark/src/main/java/org/apache/zeppelin/spark/Utils.java (hunk 1 above)
Comment (karuppayya) [inline spark/src/main/java/org/apache/zeppelin/spark/Utils.java]: done

**Comment by karuppayya** [comment]
On restrating the failed job(selenium tests), it succeeded
https://travis-ci.org/karuppayya/zeppelin/builds/246794338
Ready for reviewing

File: spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java (hunk 2 above)
Comment (felixcheung) [inline spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java]: ah nice - so this was where it came from ;)

**Comment by karuppayya** [comment]
@felixcheung 

1. Regarding the security exploitation
Spark does escapes any HTML, except for a anchor element with relative links
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ui/UIUtils.scala#L439
2. There can be multiple jobs per notebook paragraph
When a sceond paragrph with different code is executed, the corresponding paragraph code shows up in the description section

**Comment by karuppayya** [comment]
@felixcheung any suggestion on the length of text(for the short description).   
Also in case of scala, python code, when the first code line is a import , should we skip it and show the first line of actual code as short description, instead of the import statement?.

Currently, we show the entire code in the long description. This can potentially bloat the evnt logs. Should we prune it, to say last n lines?.
@felixcheung @Leemoonsoo @jongyoul

**Comment by felixcheung** [comment]
what is the goal of this? if the goal is identification, then either the last line (or last few lines) and/or with paragraph id should be sufficient.
I believe the pig interpreter has a similar approach, right @zjffdu ?

**Comment by zjffdu** [comment]
Yes, we need to truncate that. @karuppayya Could you add some comment to explain what is the jobShortText and jobLongText.

**Pull request closed without merging**